### ISSUE TITLE
Attaching to a Linux process is not yet implemented

### ISSUE THREAD

**Issue by leculver** [microsoft/clrmd#143]
I'm winding down development in this sprint of ClrMD work.  It's looking like I won't have time to handle inspecting a live process.  I'm leaving this issue open as a TODO marker and to let others know it's a known issue.

**Comment by TylerAP** [comment]
ran into this.

**Comment by leculver** [comment]
For now you should be able to use the createdump utility to drop a core dump, then load it.

https://github.com/dotnet/coreclr/blob/master/Documentation/botr/xplat-minidump-generation.md

**Comment by TylerAP** [comment]
that's exactly what I'm going to do... I just have to get createdump to work in a self-contained docker image based on runtime-deps...

**Comment by leculver** [comment]
I will have another sprint of time to do some work in late December, which is when I _plan_ to implement live process attach on Linux.  At that point I'll stop calling the Linux support "experimental"...

**Comment by TylerAP** [comment]
FYI for anyone else that is looking for createdump in containers...
It's not in microsoft/dotnet:*-runtime-deps.
In microsoft/dotnet:*-runtime / *-sdk, it's in ...
`/usr/share/dotnet/shared/Microsoft.NETCore.App/*/createdump`

**Comment by leculver** [comment]
This is a problem with some of our diagnostic tools.  Overall I'd say this probably needs attention from our diagnostics team that this doesn't work out of the box (so @mikem8361 @noahfalk).

I've worked around this locally by renaming "libcoreclrtraceptprovider.so" to something like "libcoreclrtraceptprovider.s_".  This makes it so that CLR and the dac can't find this tracepoint provider library and you avoid this error.

Sorry for the trouble here, I'm not sure what the status of fixing this problem is.

**Comment by mikem8361** [comment]
This has been fixed for 3.0 (master branch) in https://github.com/dotnet/coreclr/pull/20874.  I'm investigating getting it fixed in 2.2.

**Comment by leculver** [comment]
Awesome!

Yeah we won't be able to accept any PR that gets rid of the arcade and other scripts.  All of this goo is basically required for us to be able to produce official, signed NuGet packages.

However, if you have a working implementation of attaching to a live linux process, I'd be super happy to have that as a PR.  One thing to be aware of is the JetBrains team is looking to submit a very large cleanup PR for the project.  The draft of it is here:  https://github.com/Microsoft/clrmd/pull/151

This will almost certainly conflict with any changes you are making.  You may want to wait until that PR is wrapped up and merged, then refork.  It's up to you.

**Comment by TylerAP** [comment]
I'll rebase on that and keep sync'd. Thanks for the heads up.

**Comment by leculver** [comment]
@TylerAP Ok, Microsoft/clrmd/master is now fully up to date with all the refactoring/code changes that will be done.  Sorry for any inconvenience it caused...I wanted to set a new baseline for future development.  (Also thanks to Ruslan who put these changes together.)

**Comment by TylerAP** [comment]
I rebased earlier. I am sync'd with master. No big deal.

I am making progress. Working on generating core dumps under Linux.

```msbuild
        <Exec Command="dotnet '$(TargetPath)'" EnvironmentVariables="COMPlus_DbgMiniDumpName=$(TargetDumpFileName)_wks_mini_xplat.dmp;COMPlus_DbgEnableMiniDump=1;COMPlus_DbgMiniDumpType=1;COMPLUS_BuildFlavor=WKS;COMPlus_gcServer=0" ContinueOnError="true" ConsoleToMSBuild="true">
            <Output TaskParameter="ConsoleOutput" PropertyName="TargetOutputWksMini" />
        </Exec>

        <Exec Command="dotnet '$(TargetPath)'" EnvironmentVariables="COMPlus_DbgMiniDumpName=$(TargetDumpFileName)_wks_xplat.dmp;COMPlus_DbgEnableMiniDump=1;COMPlus_DbgMiniDumpType=4;COMPLUS_BuildFlavor=WKS;COMPlus_gcServer=0" ContinueOnError="true" ConsoleToMSBuild="true">
            <Output TaskParameter="ConsoleOutput" PropertyName="TargetOutputWks" />
        </Exec>

        <Exec Command="dotnet '$(TargetPath)'" EnvironmentVariables="COMPlus_DbgMiniDumpName=$(TargetDumpFileName)_svr_mini_xplat.dmp;COMPlus_DbgEnableMiniDump=1;COMPlus_DbgMiniDumpType=1;COMPLUS_BuildFlavor=SVR;COMPlus_gcServer=1" ContinueOnError="true" ConsoleToMSBuild="true">
            <Output TaskParameter="ConsoleOutput" PropertyName="TargetOutputSvrMini" />
        </Exec>

        <Exec Command="dotnet '$(TargetPath)'" EnvironmentVariables="COMPlus_DbgMiniDumpName=$(TargetDumpFileName)_svr_xplat.dmp;COMPlus_DbgEnableMiniDump=1;COMPlus_DbgMiniDumpType=4;COMPLUS_BuildFlavor=SVR;COMPlus_gcServer=1" ContinueOnError="true" ConsoleToMSBuild="true">
            <Output TaskParameter="ConsoleOutput" PropertyName="TargetOutputSvr" />
        </Exec>


        <Exec Command="ls -la '$(TargetDir)'" ConsoleToMSBuild="true">
            <Output TaskParameter="ConsoleOutput" PropertyName="TargetOutputDirList" />
        </Exec>

        <Error Text="$(TargetOutputDirList.Replace(';','&#10;'))" ContinueOnError="true" />

        <Error Condition="!Exists('$(TargetDumpFileName)_wks_mini_xplat.dmp')" Text="Failed to create $(TargetDumpFileName)_wks_mini_xplat.dmp'&#10;'$(TargetOutputWksMini.Replace(';','&#10;'))" />
        <Error Condition="!Exists('$(TargetDumpFileName)_wks_xplat.dmp')" Text="Failed to create $(TargetDumpFileName)_wks_xplat.dmp'&#10;'$(TargetOutputWks.Replace(';','&#10;'))" />
        <Error Condition="!Exists('$(TargetDumpFileName)_svr_mini_xplat.dmp')" Text="Failed to create $(TargetDumpFileName)_svr_mini_xplat.dmp'&#10;'$(TargetOutputSvrMini.Replace(';','&#10;'))" />
        <Error Condition="!Exists('$(TargetDumpFileName)_svr_xplat.dmp')" Text="Failed to create $(TargetDumpFileName)_svr_xplat.dmp'&#10;'$(TargetOutputSvr.Replace(';','&#10;'))" />
```

```
/app/src/TestTargets/TestTargets.csproj(57,9): warning MSB3073: The command "dotnet '/app/src/TestTargets/bin/x64/AppDomains.dll'" exited with code 134.
/app/src/TestTargets/TestTargets.csproj(61,9): warning MSB3073: The command "dotnet '/app/src/TestTargets/bin/x64/AppDomains.dll'" exited with code 134.
/app/src/TestTargets/TestTargets.csproj(65,9): warning MSB3073: The command "dotnet '/app/src/TestTargets/bin/x64/AppDomains.dll'" exited with code 134.
/app/src/TestTargets/TestTargets.csproj(69,9): warning MSB3073: The command "dotnet '/app/src/TestTargets/bin/x64/AppDomains.dll'" exited with code 134.
/app/src/TestTargets/TestTargets.csproj(78,9): warning : total 44
/app/src/TestTargets/TestTargets.csproj(78,9): warning : drwxr-xr-x 2 root root 4096 Dec  1 11:15 .
/app/src/TestTargets/TestTargets.csproj(78,9): warning : drwxr-xr-x 5 root root 4096 Dec  1 11:15 ..
/app/src/TestTargets/TestTargets.csproj(78,9): warning : -rw-r--r-- 1 root root  823 Dec  1 11:15 AppDomains.deps.json
/app/src/TestTargets/TestTargets.csproj(78,9): warning : -rw-r--r-- 1 root root 4608 Dec  1 11:15 AppDomains.dll
/app/src/TestTargets/TestTargets.csproj(78,9): warning : -rw-r--r-- 1 root root  548 Dec  1 11:15 AppDomains.pdb
/app/src/TestTargets/TestTargets.csproj(78,9): warning : -rw-r--r-- 1 root root  190 Dec  1 11:15 AppDomains.runtimeconfig.dev.json
/app/src/TestTargets/TestTargets.csproj(78,9): warning : -rw-r--r-- 1 root root  146 Dec  1 11:15 AppDomains.runtimeconfig.json
/app/src/TestTargets/TestTargets.csproj(78,9): warning : -rw-r--r-- 3 root root 4608 Dec  1 11:15 SharedLibrary.dll
/app/src/TestTargets/TestTargets.csproj(78,9): warning : -rw-r--r-- 3 root root  564 Dec  1 11:15 SharedLibrary.pdb
/app/src/TestTargets/TestTargets.csproj(80,9): error : Failed to create /app/src/TestTargets/bin/x64/AppDomains_wks_mini_xplat.dmp'
/app/src/TestTargets/TestTargets.csproj(80,9): error : 'Unhandled Exception: System.PlatformNotSupportedException: Secondary AppDomains are not supported on this platform.;at System.AppDomain.CreateDomain(String friendlyName);at Program.Main(String[] args) in /app/src/TestTargets/AppDomains.cs:line 11;ptrace(ATTACH, 72) FAILED Operation not permitted;Aborted (core dumped);Writing minidump to file /app/src/TestTargets/bin/x64/AppDomains_wks_mini_xplat.dmp
```
Turns out dispite the report of "Writing minidump to file /app/src/TestTargets/bin/x64/AppDomains_wks_mini_xplat.dmp" it does in fact not appear to write a file... I might have something misconfigured. I dunno yet.

**Comment by TylerAP** [comment]
I'm [sync'd.](https://github.com/TylerAP/clrmd/pull/2)

Like I said though, I'll part out PRs I submit back, and at some point I'm going to squash a lot of my git history and restructure it to fit your flow.

**Comment by TylerAP** [comment]
Looks like JetBrains is continuing to do more refactoring.
Should I plan to sync up with [Ruslan Isakiev's fork](https://github.com/isakiev/clrmd) again in the *near* future? I do like the look of that `.editorconfig`.

**Comment by TYoung86** [comment]
I'd hope that the windows natives could all be split out from classes that contain any non-windows (xplat/posix/linux/bsd) natives in a refactoring pass. I've done that on my branch to avoid static initialization causing `DllNotFoundException`s.

I've created an XPlatLiveDataTarget (and an XPlatDataReader, but then merged it into XPlatLiveDataTarget).
I've got initial Passive attachment started. I'm currently working through getting the dac loaded under linux, and encountering 0x80131c4f ... Got plenty of leads, working through it.

**Comment by TylerAP** [comment]
Judging by [coreclr/src/daccess/daccess.cpp lines 7251 to 7301](https://github.com/dotnet/coreclr/blob/master/src/debug/daccess/daccess.cpp#L7251-L7301) ... CORDBG_E_MISSING_DEBUGGER_EXPORTS is given when one of the calls fail in ClrDataAccess::GetDacGlobals() to the wrapped data target. So I'm going to revert any attempt to use other interfaces and keep with CLRDataCreateInstance.

**Comment by YoungCP** [comment]
clrmd-tests_1  | Build started 12/06/2018 22:39:46.
clrmd-tests_1  | Test run for /app/src/Microsoft.Diagnostics.Runtime.Tests/bin/Debug/netcoreapp2.1/Microsoft.Diagnostics.Runtime.Tests.dll(.NETCoreApp,Version=v2.1)
clrmd-tests_1  | Microsoft (R) Test Execution Command Line Tool Version 15.9.0
clrmd-tests_1  | Copyright (c) Microsoft Corporation.  All rights reserved.
clrmd-tests_1  |
clrmd-tests_1  | Starting test execution, please wait...
clrmd-tests_1  | [xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.4.0 (64-bit .NET Core 4.6.27019.06)
clrmd-tests_1  | [xUnit.net 00:00:00.56]   Discovering: Microsoft.Diagnostics.Runtime.Tests
clrmd-tests_1  | [xUnit.net 00:00:00.61]   Discovered:  Microsoft.Diagnostics.Runtime.Tests
clrmd-tests_1  | [xUnit.net 00:00:00.62]   Starting:    Microsoft.Diagnostics.Runtime.Tests
clrmd-tests_1  | [xUnit.net 00:00:00.76]     Microsoft.Diagnostics.Runtime.Tests.AttachTests.InvasiveAttachTest [SKIP]
clrmd-tests_1  | [xUnit.net 00:00:00.76]       derp
clrmd-tests_1  | [xUnit.net 00:00:01.01]     Microsoft.Diagnostics.Runtime.Tests.AttachTests.NonInvasiveAttachTest [SKIP]
clrmd-tests_1  | [xUnit.net 00:00:01.01]       derp
clrmd-tests_1  | [xUnit.net 00:00:01.02]   Finished:    Microsoft.Diagnostics.Runtime.Tests
clrmd-tests_1  | Skipped  Microsoft.Diagnostics.Runtime.Tests.AttachTests.InvasiveAttachTest
clrmd-tests_1  | Passed   Microsoft.Diagnostics.Runtime.Tests.AttachTests.PassiveAttachTest
clrmd-tests_1  | Skipped  Microsoft.Diagnostics.Runtime.Tests.AttachTests.NonInvasiveAttachTest
clrmd-tests_1  |
clrmd-tests_1  | Total tests: 3. Passed: 1. Failed: 0. Skipped: 2.
clrmd-tests_1  | Test Run Successful.
clrmd-tests_1  | Test execution time: 3.3644 Seconds
clrmd-tests_1  |
clrmd-tests_1  | Build succeeded.
clrmd-tests_1  |     0 Warning(s)
clrmd-tests_1  |     0 Error(s)
clrmd-tests_1  |
clrmd-tests_1  | Time Elapsed 00:00:05.13
clrmd_clrmd-tests_1 exited with code 0

First attach test on linux passing. Going to get invasive and non-invasive tests written and then create a new branch from main and port minimal changes. 👍  @leculver Huzzah.

**Comment by TylerAP** [comment]
(Sorry about re-post, was posting under alternate profiles by accident)

Ok, so it was my implementation of ReadMemory I needed to fix... I'm getting EPERM from process_vm_readv, classic...

<details><summary>First tests pass...</summary>

```
clrmd-tests_1  | Build started 12/06/2018 22:39:46.
clrmd-tests_1  | Test run for /app/src/Microsoft.Diagnostics.Runtime.Tests/bin/Debug/netcoreapp2.1/Microsoft.Diagnostics.Runtime.Tests.dll(.NETCoreApp,Version=v2.1)
clrmd-tests_1  | Microsoft (R) Test Execution Command Line Tool Version 15.9.0
clrmd-tests_1  | Copyright (c) Microsoft Corporation.  All rights reserved.
clrmd-tests_1  |
clrmd-tests_1  | Starting test execution, please wait...
clrmd-tests_1  | [xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.4.0 (64-bit .NET Core 4.6.27019.06)
clrmd-tests_1  | [xUnit.net 00:00:00.56]   Discovering: Microsoft.Diagnostics.Runtime.Tests
clrmd-tests_1  | [xUnit.net 00:00:00.61]   Discovered:  Microsoft.Diagnostics.Runtime.Tests
clrmd-tests_1  | [xUnit.net 00:00:00.62]   Starting:    Microsoft.Diagnostics.Runtime.Tests
clrmd-tests_1  | [xUnit.net 00:00:00.76]     Microsoft.Diagnostics.Runtime.Tests.AttachTests.InvasiveAttachTest [SKIP]
clrmd-tests_1  | [xUnit.net 00:00:00.76]       derp
clrmd-tests_1  | [xUnit.net 00:00:01.01]     Microsoft.Diagnostics.Runtime.Tests.AttachTests.NonInvasiveAttachTest [SKIP]
clrmd-tests_1  | [xUnit.net 00:00:01.01]       derp
clrmd-tests_1  | [xUnit.net 00:00:01.02]   Finished:    Microsoft.Diagnostics.Runtime.Tests
clrmd-tests_1  | Skipped  Microsoft.Diagnostics.Runtime.Tests.AttachTests.InvasiveAttachTest
clrmd-tests_1  | Passed   Microsoft.Diagnostics.Runtime.Tests.AttachTests.PassiveAttachTest
clrmd-tests_1  | Skipped  Microsoft.Diagnostics.Runtime.Tests.AttachTests.NonInvasiveAttachTest
clrmd-tests_1  |
clrmd-tests_1  | Total tests: 3. Passed: 1. Failed: 0. Skipped: 2.
clrmd-tests_1  | Test Run Successful.
clrmd-tests_1  | Test execution time: 3.3644 Seconds
clrmd-tests_1  |
clrmd-tests_1  | Build succeeded.
clrmd-tests_1  |     0 Warning(s)
clrmd-tests_1  |     0 Error(s)
clrmd-tests_1  |
clrmd-tests_1  | Time Elapsed 00:00:05.13
clrmd_clrmd-tests_1 exited with code 0
```

</details>


First attach test on linux passing. Going to get invasive and non-invasive tests written and then create a new branch from main and port minimal changes. 👍 Huzzah. 👍

**Comment by TylerAP** [comment]
Enumerated handles on an attached process under Linux. :godmode: 
This satisfies for my needs, the thread implementation will be a nice to have.
Hopefully I will get some more time to spend on this after hours.

https://github.com/TylerAP/clrmd/issues/1#issuecomment-445091021

**Comment by adamsitnik** [comment]
I just wanted to say that it would be awesome to have it, it would unblock BenchmarkDotNet to make our disassembler cross-platform https://github.com/dotnet/BenchmarkDotNet/pull/1015

**Comment by akrock** [comment]
We have used CLR MD in the past to live attach to our own process and dump running threads to assist with identifying deadlocks, and other unexpected processing slowdowns for batch worker processes in full framework. 

As we are now migrating some of our batch processes into NET CORE and linux docker containers we would be very interested in having this functionality working under linux.

**Comment by loop-evgeny** [comment]
Any update on this? It would be great to have!

**Comment by leculver** [comment]
@akrock @loop-evgeny Live data reading was added in pull https://github.com/microsoft/clrmd/pull/263.  Also Mike made some fixes in pull https://github.com/microsoft/clrmd/pull/276.  I have not had the chance to test it but I've been assured it works, so I will close this issue.  If you find problems with it please feel free to open new issues about the specific problems.

**Comment by loop-evgeny** [comment]
Yes, it works now! Thanks so much!
